File: backend/app/Routes/regions.py
# backend/app/Routes/regions.py
from flask import Blueprint, request, jsonify
from datetime import date as Date
from app.extensions import db
from app.Models.Region import Region
from app.Models.Fire.fires_daily import FiresDaily
from app.Models.Weather.weather_daily_regional import WeatherDailyRegional
from app.Models.Fire.firms_viirs import FirmsVIIRS

regions_bp = Blueprint("regions", __name__, url_prefix="/api/regions")


def _to_date(s: str) -> Date:
    """Parse YYYY-MM-DD string to date."""
    try:
        return Date.fromisoformat(s)
    except ValueError:
        raise ValueError(f"Invalid date format: {s}. Expected YYYY-MM-DD")


@regions_bp.route("", methods=["GET"])
def list_regions():
    """
    GET /api/regions
    Returns list of all regions with metadata.
    """
    regions = Region.query.order_by(Region.slug).all()

    data = [
        {
            "slug": r.slug,
            "name": r.name,
            "center": {"lat": r.center_lat, "lon": r.center_lon},
            "bbox": {
                "west": r.bbox_w,
                "south": r.bbox_s,
                "east": r.bbox_e,
                "north": r.bbox_n,
            },
        }
        for r in regions
    ]

    return jsonify({"count": len(data), "regions": data})


@regions_bp.route("/<slug>/fires", methods=["GET"])
def get_region_fires(slug: str):
    """
    GET /api/regions/<slug>/fires?start=YYYY-MM-DD&end=YYYY-MM-DD
    Returns daily fire aggregates for a region.
    """
    start_str = request.args.get("start")
    end_str = request.args.get("end")

    if not (start_str and end_str):
        return jsonify({"error": "start and end query params are required (YYYY-MM-DD)"}), 400

    try:
        start = _to_date(start_str)
        end = _to_date(end_str)
    except ValueError as e:
        return jsonify({"error": str(e)}), 400

    # Find region
    region = Region.query.filter_by(slug=slug).first()
    if not region:
        return jsonify({"error": f"Region not found: {slug}"}), 404

    # Query fires_daily
    rows = (
        FiresDaily.query
        .filter(FiresDaily.region_id == region.id)
        .filter(FiresDaily.acq_date >= start, FiresDaily.acq_date <= end)
        .order_by(FiresDaily.acq_date)
        .all()
    )

    data = [
        {
            "date": r.acq_date.isoformat(),
            "fire_count": r.fire_count,
            "avg_brightness": r.avg_brightness,
            "avg_confidence": r.avg_confidence,
            "avg_frp": r.avg_frp,
            "label": r.label,
        }
        for r in rows
    ]

    return jsonify({
        "region": slug,
        "start": start.isoformat(),
        "end": end.isoformat(),
        "count": len(data),
        "fires": data,
    })


@regions_bp.route("/<slug>/weather", methods=["GET"])
def get_region_weather(slug: str):
    """
    GET /api/regions/<slug>/weather?start=YYYY-MM-DD&end=YYYY-MM-DD
    Returns daily weather for a region.
    """
    start_str = request.args.get("start")
    end_str = request.args.get("end")

    if not (start_str and end_str):
        return jsonify({"error": "start and end query params are required (YYYY-MM-DD)"}), 400

    try:
        start = _to_date(start_str)
        end = _to_date(end_str)
    except ValueError as e:
        return jsonify({"error": str(e)}), 400

    # Find region
    region = Region.query.filter_by(slug=slug).first()
    if not region:
        return jsonify({"error": f"Region not found: {slug}"}), 404

    # Query weather_daily_regional
    rows = (
        WeatherDailyRegional.query
        .filter(WeatherDailyRegional.region_id == region.id)
        .filter(WeatherDailyRegional.date >= start, WeatherDailyRegional.date <= end)
        .order_by(WeatherDailyRegional.date)
        .all()
    )

    data = [
        {
            "date": r.date.isoformat(),
            "tempmax": r.tempmax,
            "tempmin": r.tempmin,
            "humidity": r.humidity,
            "windspeed": r.windspeed,
            "precip": r.precip,
        }
        for r in rows
    ]

    return jsonify({
        "region": slug,
        "start": start.isoformat(),
        "end": end.isoformat(),
        "count": len(data),
        "weather": data,
    })


@regions_bp.route("/<slug>/map-firms", methods=["GET"])
def get_region_map_firms(slug: str):
    """
    GET /api/regions/<slug>/map-firms?start=YYYY-MM-DD&end=YYYY-MM-DD&min_conf=0&max=5000
    Convenience endpoint: uses region bbox to query firms_viirs for map display.
    """
    start_str = request.args.get("start")
    end_str = request.args.get("end")
    min_conf = float(request.args.get("min_conf", "0"))
    limit = int(request.args.get("max", "5000"))

    if not (start_str and end_str):
        return jsonify({"error": "start and end query params are required (YYYY-MM-DD)"}), 400

    try:
        start = _to_date(start_str)
        end = _to_date(end_str)
    except ValueError as e:
        return jsonify({"error": str(e)}), 400

    # Find region
    region = Region.query.filter_by(slug=slug).first()
    if not region:
        return jsonify({"error": f"Region not found: {slug}"}), 404

    # Query firms_viirs using region bbox
    q = (
        db.session.query(FirmsVIIRS)
        .filter(FirmsVIIRS.acq_date >= start, FirmsVIIRS.acq_date <= end)
        .filter(FirmsVIIRS.latitude.between(region.bbox_s, region.bbox_n))
        .filter(FirmsVIIRS.longitude.between(region.bbox_w, region.bbox_e))
    )

    if min_conf > 0:
        q = q.filter(FirmsVIIRS.confidence >= min_conf)

    rows = q.order_by(FirmsVIIRS.acq_date.desc()).limit(limit).all()

    data = [
        {
            "lat": r.latitude,
            "lon": r.longitude,
            "date": r.acq_date.isoformat(),
            "conf": r.confidence,
            "sat": r.satellite,
            "dn": r.daynight,
        }
        for r in rows
    ]

    return jsonify({
        "region": slug,
        "count": len(data),
        "items": data,
    })

File: frontend/src/api/regions.ts
// src/api/regions.ts
/// <reference types="vite/client" />

export type RegionInfo = {
  slug: string;
  name: string;
  center: { lat: number; lon: number };
  bbox: {
    west: number;
    south: number;
    east: number;
    north: number;
  };
};

export type RegionsResponse = {
  count: number;
  regions: RegionInfo[];
};

export type FireDailyRow = {
  date: string; // YYYY-MM-DD
  fire_count: number;
  avg_brightness: number | null;
  avg_confidence: number | null;
  avg_frp: number | null;
  label: number;
};

export type WeatherDailyRow = {
  date: string; // YYYY-MM-DD
  tempmax: number | null;
  tempmin: number | null;
  humidity: number | null;
  windspeed: number | null;
  precip: number | null;
};

export type RegionFiresResponse = {
  region: string;
  start: string;
  end: string;
  count: number;
  fires: FireDailyRow[];
};

export type RegionWeatherResponse = {
  region: string;
  start: string;
  end: string;
  count: number;
  weather: WeatherDailyRow[];
};

export type RegionMapFirmsItem = {
  lat: number;
  lon: number;
  date: string;
  conf: number | null;
  sat: string | null;
  dn: string | null;
};

export type RegionMapFirmsResponse = {
  region: string;
  count: number;
  items: RegionMapFirmsItem[];
};

/* ========== API base configuration ========== */

declare global {
  interface Window {
    WILDSIGHT_API_BASE?: string;
  }
}

const envBase = import.meta.env.VITE_API_BASE;
const runtimeBase =
  typeof window !== "undefined" ? window.WILDSIGHT_API_BASE : undefined;

const API_BASE = (runtimeBase ?? envBase ?? "http://localhost:5005").replace(
  /\/+$/,
  ""
);

/* ========== JSON fetch helper ========== */

async function fetchJSON(url: string, init?: RequestInit) {
  const res = await fetch(url, init);
  const text = await res.text();

  let data: any;
  try {
    data = text ? JSON.parse(text) : {};
  } catch {
    throw new Error(`Non-JSON response from backend: ${text.slice(0, 200)}`);
  }

  if (!res.ok) {
    throw new Error(data?.error ?? `HTTP ${res.status}`);
  }

  return data;
}

/* ========== Public API functions ========== */

/**
 * GET /api/regions
 * Fetches list of all available regions.
 */
export async function fetchRegions(): Promise<RegionsResponse> {
  const url = `${API_BASE}/api/regions`;
  return fetchJSON(url);
}

/**
 * GET /api/regions/<slug>/fires?start=YYYY-MM-DD&end=YYYY-MM-DD
 * Fetches daily fire aggregates for a region.
 */
export async function fetchRegionFires(
  slug: string,
  start: string,
  end: string
): Promise<RegionFiresResponse> {
  const params = new URLSearchParams({ start, end });
  const url = `${API_BASE}/api/regions/${slug}/fires?${params}`;
  return fetchJSON(url);
}

/**
 * GET /api/regions/<slug>/weather?start=YYYY-MM-DD&end=YYYY-MM-DD
 * Fetches daily weather for a region.
 */
export async function fetchRegionWeather(
  slug: string,
  start: string,
  end: string
): Promise<RegionWeatherResponse> {
  const params = new URLSearchParams({ start, end });
  const url = `${API_BASE}/api/regions/${slug}/weather?${params}`;
  return fetchJSON(url);
}

/**
 * GET /api/regions/<slug>/map-firms?start=YYYY-MM-DD&end=YYYY-MM-DD&min_conf=0&max=5000
 * Fetches FIRMS detections for a region (convenience endpoint using region bbox).
 */
export async function fetchRegionMapFirms(
  slug: string,
  start: string,
  end: string,
  options?: { minConf?: number; max?: number }
): Promise<RegionMapFirmsResponse> {
  const params = new URLSearchParams({ start, end });
  if (options?.minConf !== undefined) params.set("min_conf", String(options.minConf));
  if (options?.max !== undefined) params.set("max", String(options.max));

  const url = `${API_BASE}/api/regions/${slug}/map-firms?${params}`;
  return fetchJSON(url);
}
